### ISSUE TITLE
Thread-local memory in actions

### ISSUE THREAD

**Issue by nilsvu** [sxs-collaboration/spectre#5120]
We often allocate memory in actions and throw it away at the end of the action to keep memory pressure down (we can't have all elements keep memory buffers for temporaries around in the DataBox). It would be nice to eliminate these frequent memory allocations, and it seems possible because elements on one thread only execute actions sequentially. Therefore we could have thread-local memory that every element on the thread uses. Elements may have to resize the memory to make it bigger, and smaller elements just use a chunk of it.

Questions:

1. Would this be worth the effort? We have to profile.
2. What would be reasonable implementations?

**Comment by nilsdeppe** [comment]
1. Currently profiling GH when using jemalloc (I haven't looked at not using jemalloc) the overhead is actually from _copying_ data, not from the allocation itself. In particular, sending boundary data costs ~10% of runtime based on cycles with hpctoolkit. (De)Allocations together were around 5%, so higher than we want for what is essentially "useless work", but a lot of that is attached to the boundary data communication, so finding a solution for reusing boundary data is I think most pressing.
2. One option, e.g. for inside an action is `static thread_local` that we resize if it's too small. This would work for the `ComputeTimeDerivative` action, but this single allocation in a time step is also not currently showing up in profilers. Another way to handle this would be some thread-local global state using [std::pmr::](https://en.cppreference.com/w/cpp/memory/unsynchronized_pool_resource) code. The PMR features are actually very nice and designed specifically for dealing with custom allocation, memory buffers, memory pools, etc. There are both thread-local and thread-shared pools.

**Comment by nilsvu** [comment]
Is it possible that memory allocations on different threads contend for access to the memory subsystem, and will this show up in profiling?

**Comment by nilsdeppe** [comment]
Yes, my experience in the past has been that if you profile 15% of time in malloc/free, and you drop it to 7%, you will actually realize a 30% reduction in runtime. Now, there has to be a lower limit to this where the system calls no longer overlap and therefore the measured percentage is "exactly" what it is, but I don't know if this is 5% of runtime, 1%, 0.001%, or exactly 0%.
